hacks/sphere.man
.TH XScreenSaver 1 "27-May-97" "X Version 11"
.SH NAME
sphere \- draws shaded spheres
.SH SYNOPSIS
.B sphere
[\-display \fIhost:display.screen\fP] [\-foreground \fIcolor\fP] [\-background \fIcolor\fP] [\-window] [\-root] [\-mono] [\-install] [\-visual \fIvisual\fP] [\-ncolors \fIinteger\fP] [\-delay \fImicroseconds\fP]

[\-fps]
.SH DESCRIPTION
The \fIsphere\fP program draws shaded spheres.
.SH OPTIONS
.I sphere
accepts the following options:
.TP 8
.B \-window
Draw on a newly-created window.  This is the default.
.TP 8
.B \-root
Draw on the root window.
.TP 8
.B \-mono 
If on a color display, pretend we're on a monochrome display.
.TP 8
.B \-install
Install a private colormap for the window.
.TP 8
.B \-visual \fIvisual\fP
Specify which visual to use.  Legal values are the name of a visual class,
or the id number (decimal or hex) of a specific visual.
.TP 8
.B \-ncolors \fIinteger\fP
How many colors should be used (if possible).  Default 64.
The colors are chosen randomly.
.TP 8
.B \-fps
Display the current frame rate and CPU load.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH SEE ALSO
.BR X (1),
.BR xscreensaver (1),
.BR xlock (1)
.SH COPYRIGHT
Copyright \(co 1988 by Sun Microsystems, Inc.

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation. 
.SH AUTHOR
Sun Microsystems, Inc, 1988.

Ability to run standalone or with \fIxscreensaver\fP added by 
Jamie Zawinski <jwz@jwz.org>, 27-May-97.


po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from
	Guillermo Movia <gmovia@hotpop.com> and converted to UTF-8

2002-04-29  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* wa.po: Updated Walloon file

2002-04-27  Duarte Loreto <happyguy_pt@hotmail.com>

        * pt.po: Added Portuguese translation.

2002-04-26  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

        * et.po: Updated Estonian translation.

2002-04-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-04-21  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmÃ¥l) translation.


hacks/sphere.c
/* -*- Mode: C; tab-width: 4 -*- */
/* sphere --- a bunch of shaded spheres */

#if 0
static const char sccsid[] = "@(#)sphere.c	5.00 2000/11/01 xlockmore";
#endif

/*-
 * Copyright (c) 1988 by Sun Microsystems
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright notice appear in all copies and that
 * both that copyright notice and this permission notice appear in
 * supporting documentation.
 *
 * This file is provided AS IS with no warranties of any kind.  The author
 * shall have no liability with respect to the infringement of copyrights,
 * trade secrets or any patents by this file or any part thereof.  In no
 * event will the author be liable for any lost revenue or profits or
 * other special, indirect and consequential damages.
 *
 * Revision History:
 * 01-Nov-2000: Allocation checks
 * 30-May-1997: <jwz@jwz.org> made it go vertically as well as horizontally.
 * 27-May-1997: <jwz@jwz.org> turned into a standalone program.
 * 02-Sep-1993: xlock version David Bagley <bagleyd@tux.org>
 * 1988: Revised to use SunView canvas instead of gfxsw Sun Microsystems
 * 1982: Orignal Algorithm Tom Duff Lucasfilm Ltd.
 */

/*-
 * original copyright
 * **************************************************************************
 * Copyright 1988 by Sun Microsystems, Inc. Mountain View, CA.
 *
 * All Rights Reserved
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the names of Sun or MIT not be used in advertising
 * or publicity pertaining to distribution of the software without specific
 * prior written permission. Sun and M.I.T. make no representations about the
 * suitability of this software for any purpose. It is provided "as is"
 * without any express or implied warranty.
 *
 * SUN DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 * IN NO EVENT SHALL SUN BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
 * ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 * SOFTWARE.
 * ***************************************************************************
 */

#ifdef STANDALONE
# define MODE_sphere
#define DEFAULTS	"*delay: 20000 \n" \
					"*cycles: 20 \n" \
					"*size: 0 \n" \
					"*ncolors: 64 \n" \
					"*fpsSolid: true \n" \

# define BRIGHT_COLORS
# define reshape_sphere 0
# define sphere_handle_event 0
# include "xlockmore.h"		/* from the xscreensaver distribution */
#else /* !STANDALONE */
# include "xlock.h"		/* from the xlockmore distribution */
#endif /* !STANDALONE */

#ifdef MODE_sphere

ENTRYPOINT ModeSpecOpt sphere_opts =
{0, (XrmOptionDescRec *) NULL, 0, (argtype *) NULL, (OptionStruct *) NULL};

#ifdef USE_MODULES
ModStruct   sphere_description =
{"sphere", "init_sphere", "draw_sphere", "release_sphere",
 "refresh_sphere", "init_sphere", (char *) NULL, &sphere_opts,
 5000, 1, 20, 0, 64, 1.0, "",
 "Shows a bunch of shaded spheres", 0, NULL};

#endif

/*-
 * (NX, NY, NZ) is the light source vector -- length should be 100
 */
#define NX 48
#define NY (-36)
#define NZ 80
#define NR 100
#define SQRT(a) ((int)sqrt((double)(a)))

typedef struct {
	int         width, height;
	int         radius;
	int         x0;		/* x center */
	int         y0;		/* y center */
	int         color;
	int         x, y;
	int         dirx, diry;
	int         shadowx, shadowy;
	int         maxx, maxy;
	XPoint     *points;
} spherestruct;

static spherestruct *spheres = (spherestruct *) NULL;

ENTRYPOINT void
init_sphere(ModeInfo * mi)
{
	spherestruct *sp;

	MI_INIT (mi, spheres, 0);
	sp = &spheres[MI_SCREEN(mi)];

	if (sp->points != NULL) {
		(void) free((void *) sp->points);
		sp->points = (XPoint *) NULL;
	}
	sp->width = MAX(MI_WIDTH(mi), 4);
	sp->height = MAX(MI_HEIGHT(mi), 4);
	if ((sp->points = (XPoint *) malloc(MIN(sp->width, sp->height) *
			 sizeof (XPoint))) == NULL) {
		return;
	}

	MI_CLEARWINDOW(mi);

	sp->dirx = 1;
	sp->x = sp->radius;
	sp->shadowx = (LRAND() & 1) ? 1 : -1;
	sp->shadowy = (LRAND() & 1) ? 1 : -1;
}

ENTRYPOINT void
draw_sphere(ModeInfo * mi)
{
	Display    *display = MI_DISPLAY(mi);
	GC          gc = MI_GC(mi);
	int         sqrd, nd;
	register int minx = 0, maxx = 0, miny = 0, maxy = 0, npts = 0;
	spherestruct *sp;

	if (spheres == NULL)
		return;
	sp = &spheres[MI_SCREEN(mi)];
	if (sp->points == NULL)
		return;

	MI_IS_DRAWN(mi) = True;
	if ((sp->dirx && ABS(sp->x) >= sp->radius) ||
	    (sp->diry && ABS(sp->y) >= sp->radius)) {
		sp->radius = NRAND(MIN(sp->width / 2, sp->height / 2) - 1) + 1;

		if (LRAND() & 1) {
			sp->dirx = (int) (LRAND() & 1) * 2 - 1;
			sp->diry = 0;
		} else {
			sp->dirx = 0;
			sp->diry = (int) (LRAND() & 1) * 2 - 1;
		}
		sp->x0 = NRAND(sp->width);
		sp->y0 = NRAND(sp->height);

		sp->x = -sp->radius * sp->dirx;
		sp->y = -sp->radius * sp->diry;

		if (MI_NPIXELS(mi) > 2)
			sp->color = NRAND(MI_NPIXELS(mi));
	}
	if (sp->dirx == 1) {
		if (sp->x0 + sp->x < 0)
			sp->x = -sp->x0;
	} else if (sp->dirx == -1) {
		if (sp->x0 + sp->x >= sp->width)
			sp->x = sp->width - sp->x0 - 1;
	}
	if (sp->diry == 1) {
		if (sp->y0 + sp->y < 0)
			sp->y = -sp->y0;
	} else if (sp->diry == -1) {
		if (sp->y0 + sp->y >= sp->height)
			sp->y = sp->height - sp->y0 - 1;
	}
	if (sp->dirx) {
		sp->maxy = SQRT(sp->radius * sp->radius - sp->x * sp->x);
		miny = -sp->maxy;
		if (sp->y0 - sp->maxy < 0)
			miny = -sp->y0;
		maxy = sp->maxy;
	}
	if (sp->diry) {
		sp->maxx = SQRT(sp->radius * sp->radius - sp->y * sp->y);
		minx = -sp->maxx;
		if (sp->x0 - sp->maxx < 0)
			minx = -sp->x0;
		maxx = sp->maxx;
	}
	if (sp->dirx) {
		if (sp->y0 + sp->maxy >= sp->height)
			maxy = sp->height - sp->y0;
	}
	if (sp->diry) {
		if (sp->x0 + sp->maxx >= sp->width)
			maxx = sp->width - sp->x0;
	}
	XSetForeground(display, gc, MI_BLACK_PIXEL(mi));

	if (sp->dirx)
		XDrawLine(display, MI_WINDOW(mi), gc,
		sp->x0 + sp->x, sp->y0 + miny, sp->x0 + sp->x, sp->y0 + maxy);
	if (sp->diry)
		XDrawLine(display, MI_WINDOW(mi), gc,
		sp->x0 + minx, sp->y0 + sp->y, sp->x0 + maxx, sp->y0 + sp->y);

	if (MI_NPIXELS(mi) > 2)
		XSetForeground(display, gc, MI_PIXEL(mi, sp->color));
	else
		XSetForeground(display, gc, MI_WHITE_PIXEL(mi));

	if (sp->dirx) {
		sqrd = sp->radius * sp->radius - sp->x * sp->x;
		nd = NX * sp->shadowx * sp->x;
		for (sp->y = miny; sp->y <= maxy; sp->y++)
			if ((NRAND(sp->radius * NR)) <= nd + NY * sp->shadowy * sp->y +
			    NZ * SQRT(sqrd - sp->y * sp->y)) {
				sp->points[npts].x = sp->x + sp->x0;
				sp->points[npts].y = sp->y + sp->y0;
				npts++;
			}
	}
	if (sp->diry) {
		sqrd = sp->radius * sp->radius - sp->y * sp->y;
		nd = NY * sp->shadowy * sp->y;
		for (sp->x = minx; sp->x <= maxx; sp->x++)
			if ((NRAND(sp->radius * NR)) <= NX * sp->shadowx * sp->x + nd +
			    NZ * SQRT(sqrd - sp->x * sp->x)) {
				sp->points[npts].x = sp->x + sp->x0;
				sp->points[npts].y = sp->y + sp->y0;
				npts++;
			}
	}
	XDrawPoints(display, MI_WINDOW(mi), gc, sp->points, npts, CoordModeOrigin);
	if (sp->dirx == 1) {
		sp->x++;
		if (sp->x0 + sp->x >= sp->width)
			sp->x = sp->radius;
	} else if (sp->dirx == -1) {
		sp->x--;
		if (sp->x0 + sp->x < 0)
			sp->x = -sp->radius;
	}
	if (sp->diry == 1) {
		sp->y++;
		if (sp->y0 + sp->y >= sp->height)
			sp->y = sp->radius;
	} else if (sp->diry == -1) {
		sp->y--;
		if (sp->y0 + sp->y < 0)
			sp->y = -sp->radius;
	}
}

ENTRYPOINT void
release_sphere(ModeInfo * mi)
{
	if (spheres != NULL) {
		int         screen;

		for (screen = 0; screen < MI_NUM_SCREENS(mi); screen++) {
			spherestruct *sp = &spheres[screen];

			if (sp->points) {
				(void) free((void *) sp->points);
				/* sp->points = NULL; */
			}
		}
		(void) free((void *) spheres);
		spheres = (spherestruct *) NULL;
	}
}

ENTRYPOINT void
refresh_sphere(ModeInfo * mi)
{
	spherestruct *sp;

	if (spheres == NULL)
		return;
	sp = &spheres[MI_SCREEN(mi)];

	MI_CLEARWINDOW(mi);

	sp->x = -sp->radius;
}

XSCREENSAVER_MODULE ("Sphere", sphere)

#endif /* MODE_sphere */
